lib/dialects/mssql/index.js

```diff
@@ -217,8 +217,11 @@ Object.assign(Client_MSSQL.prototype, {
   // Get a raw connection, called by the `pool` whenever a new
   // connection needs to be added to the pool.
   acquireRawConnection() {
-    return new Bluebird((resolver, rejecter) => {
-      const settings = Object.assign({}, this.connectionSettings);
+    const client = this;
+    return new Bluebird(async (resolver, rejecter) => {
+      const settings = await client
+        .connectionConfigProvider()
```

Comment:
So in this line we are still always initializing a new ConnectionConfigProvider, meaning we will resolve credentials again on each connection.
Can we adjust `connectionConfigProvider` method to lazy-instantiate and not reset each time?

Reply:
When a new `knex` instance defined, a new `client` is attached and a new `ConnectionConfoProvider` as well and once we call resolve connection, for the first time it will execute the function if the connection is a function then it will mutate the state of the client and set the connection to the resolved one.

I tried it on my machine and the connection is resolved only one time and for the 2nd time it loads the config from 
```
async resolveConnectionConfig() {
    if (this.resolvedConnectionSettings) return this.resolvedConnectionSettings;
```
https://github.com/knex/knex/pull/3403/files#diff-cbc1ae6887774cf34c8d4bc73dece12fR10

> Can we adjust connectionConfigProvider() method to lazy-instantiate and not reset each time?

can you give some examples how I can solve this?

Reply:
I think you are seeing connection being resolved only once because it is being reused from the connection pool and not recreated on every request - for new connections it is likely that same `acquireRawConnection` will be called again and you will have to resolve credentials again.

I'll take a swing at the branch sometime this week to see how to improve the situation.

Reply:
Easy way to check with multiple connections is to create transactions and never commit / rollback them. In that way the old connections cannot be reused from pool.

Reply:
Guys, sorry for coming in late in the process, but I noticed this PR when I was considering opening a similar issue, with the use case in mind being support of [Amazon RDS IAM authentication](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.Connecting.AWSCLI.PostgreSQL.html).

For this use case though, the resolved configuration can't be cached indefinitely, because the password needs to be changed every 15 minutes.

Can we please add an optional field to the returned configuration object, determining whether it should be cached or not?

```javascript
// Something like:
{
    host : '127.0.0.1',
    user : 'your_database_user',
    password : 'your_database_password',
    database : 'myapp_test'
    cacheResolvedConfig: true
}

// Or even more powerful, so we'll be able to cache for exactly 15 minutes:
{
    host : '127.0.0.1',
    user : 'your_database_user',
    password : 'your_database_password',
    database : 'myapp_test'
    cacheResolvedConfig: (config) => config.awsTokenEnd >= Date.now()
}
```

What do you think?